Viewport Animation: Adding flyTo style animation support.

1chandu opened the pull request:
- Adding a utility method to achieve flyTo style animation.
- Copy react-map-gl's `viewport-animation` example and update to test flyTo style animation.
- Add new props to support map interaction and loop style animations.

Next steps:
   - documentation.
   - update `viewport-animation` to test all combinations of animation props.

ibgreen commented on examples/wip/data/cities.json:
```diff
@@ -0,0 +1,8 @@
+[
```
Normally we put data files in separate repo, to avoid bloating this repo.

But this file is so small that you can keep it here. I'd suggest put it directly in the example folder, we want it to be easy to copy an example folder without having dependencies on other folders.

If it is used by more than one example, just copy it into both

shaojingli replied:
@1chandu it seems that this file is still in /wip/data/. is there a reason why @ibgreen's comment is not accepted?

1chandu replied:
@shaojingli , as mentioned in my comments, I am going to address all example related comments in following PRs, I had to merge this as per @ibgreen request to unblock him.

ibgreen commented on examples/wip/viewport-animation-3d-heatmap/app.js:
```diff
@@ -40,7 +40,7 @@ class Root extends Component {
     this._resize();
 
     // TODO: this is to just simulate viwport prop change and test animation.
```
Typo `viwport`

shaojingli replied:
ditto

1chandu replied:
will cleanup examples in following PRs.

ibgreen commented on examples/wip/viewport-animation-3d-heatmap/app.js:
```diff
@@ -61,7 +61,7 @@ class Root extends Component {
     const newViewport = {};
     newViewport.pitch = (this.state.viewport.pitch === 0) ? 60.0 : 0.0;
-    newViewport.bearing = (this.state.viewport.bearing === 0) ? -60.0 : 0.0;
+    newViewport.bearing = (this.state.viewport.bearing === 0) ? -90.0 : 0.0;
```
Nit: Magic numbers, should we define constants? E.g.:
```
const VIEWPORT1 = { ... }; 
...
```

ibgreen commented on examples/wip/viewport-animation-3d-heatmap/app.js:
```diff
@@ -73,7 +73,7 @@ class Root extends Component {
         {...viewport}
         onViewportChange={this._onViewportChange.bind(this)}
-        animateViewport={true}>
+        viewportAnimationDuration={2000}>
```
That is a long prop name. It will of course be discussed in API audit, but I suspect `duration` or `animationDuration` would work well?

ibgreen commented on examples/wip/viewport-animation/src/app.js:
```diff
@@ -0,0 +1,101 @@
+const ENABLE_ZOOM_CHANGE = false;
+const RANDOMIZE_ZOOM = false;
+
+if (!token) {
```
Nit: This should be handled by `react-map-gl`. Test if that is the case and delete here?

ibgreen commented on examples/wip/viewport-animation/src/app.js:
```diff
@@ -0,0 +1,101 @@
+        {...viewport}
+        onViewportChange={this._onViewportChange.bind(this)}
+        animateViewport={true}
+        viewportAnimationFunc={viewportFlyToAnimation}
```
`viewportAnimationFunc`

Nit: I don't like abbreviated identifiers in general, and in a code base that does not use abbreviated identifiers anywhere else it really sticks out.

ibgreen commented on examples/wip/viewport-animation/webpack.config.js:
```diff
@@ -0,0 +1,59 @@
+  resolve: {
+    alias: {
+      // From mapbox-gl-js README. Required for non-browserify bundlers (e.g. webpack):
+      // 'mapbox-gl$': resolve('./node_modules/mapbox-gl/dist/mapbox-gl.js')
```
Nit: I realize this is copied code, but can we clean up the comments?

ibgreen commented on src/experimental/react/controllers/animation-map-controller.js:
```diff
@@ -48,10 +50,11 @@ const propTypes = {
 
-  // animate viewport change.
-  animateViewport: PropTypes.bool,
+  /** Viewport animation **/
```
Nit: Use inline comments? Same line, to keep this from being too verbose?
```
  animateViewport: PropTypes.bool, // animate viewport change.
```

ibgreen commented on src/experimental/react/controllers/animation-map-controller.js:
```diff
@@ -83,8 +86,8 @@ const getDefaultCursor = ({isDragging}) => isDragging ? CURSOR.GRABBING : CURSOR
   onViewportChange: null,
-  animateViewport: false,
   viewportAnimationDuration: VIEWPORT_ANIMATION_DURATION,
+  viewportAnimationFunc: viewportLinearAnimation,
```
Probably for a follow-up PR: Can this be generalized? Does it need to be called `viewportLinearAnimation`, or can it just be `linearAnimation`?

ibgreen commented on src/experimental/react/controllers/animation-map-controller.js:
```diff
@@ -193,13 +193,12 @@ export default class AnimationMapController extends PureComponent {
+    const animatedViewport = this.props.viewportAnimationFunc(
+      this.state.animationStartViewport,
+      this.state.animationEndViewport,
+      t
```
Let's avoid abbreviated identifiers. Typing out `time` will increase readability and the minifier will reduce it back to `t` in production code.

ibgreen commented on src/experimental/react/controllers/animation-map-controller.js:
```diff
@@ -163,21 +165,19 @@ export default class AnimationMapController extends PureComponent {
+      if (this._didViewportAnimatePropChanged(startViewport, endViewport)) {
+        const animationInterval = this._createAnimationInterval();
+        this.setState({
+          animationT: 0.0,
```
Let's not abbreviate identifiers, unless we have a strong reason. Why is `animationT` better than `animationTime`?

ibgreen commented on src/experimental/react/controllers/viewport-animation-utils.js:
```diff
@@ -0,0 +1,115 @@
+/* eslint max-statements: ["error", 50] */
```
Since this is not a react component, this should go into `src/controllers` instead of `src/react/controllers`

ibgreen commented on src/experimental/react/controllers/viewport-animation-utils.js:
```diff
@@ -0,0 +1,115 @@
+ * react/redux architecture.
+ * mapbox-gl-js flyTo : https://www.mapbox.com/mapbox-gl-js/api/#map#flyto.
+ * It implements “Smooth and efficient zooming and panning.” algorithm by
+ * "Jarke J. van Wijk and Wim A.A. Nuij"
```
Link to paper?

Pessimistress commented on examples/wip/viewport-animation/src/app.js:
```diff
@@ -0,0 +1,101 @@
+import React, {Component} from 'react';
+import {StaticMap} from 'react-map-gl';
+import {experimental} from 'deck.gl';
+import TWEEN from 'tween.js';
```
Looks like tweenjs is no longer a dependency - remove?

Pessimistress commented on examples/wip/viewport-animation/src/app.js:
```diff
@@ -0,0 +1,101 @@
+const {AnimationMapController, viewportFlyToAnimation} = experimental;
+const token = process.env.MapboxAccessToken; // eslint-disable-line
+const ENABLE_ZOOM_CHANGE = false;
+const RANDOMIZE_ZOOM = false;
```
I assume these are useful for debugging, but we should keep examples minimal and easy to read.

1chandu commented:
I have addressed most of the comments except example cleanup, plan to merge this to unblock others, I will address remaining comments in next PR. All the animation features work fine.
